core_java/twitter/src/main/java/ca/jrvs/apps/twitter/DAO/TwitterService.java

```diff
@@ -0,0 +1,129 @@
+package ca.jrvs.apps.twitter.DAO;
+
+import ca.jrvs.apps.twitter.model.Coordinates;
+import ca.jrvs.apps.twitter.model.Tweet;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TwitterService implements Service {
+
+  private CrdDao dao;
+
+  //@Autowired
+  public TwitterService(CrdDao dao) {
+    this.dao = dao;
+  }
+
+  @Override
+  public Tweet postTweet(Tweet tweet) {
+    //get Text
+    //get long/lat
+    if (tweet == null) {
+      throw new IllegalArgumentException("Value of Tweet is null");
+    }
+    Coordinates coordinates = tweet.getCoordinates();
+    int textLength = tweet.getText().length();
+    if (textLength > 140) {
+      throw new IllegalArgumentException("Text over 140 characters");
+    }
+
+    if (coordinates != null) {
+      float lon = coordinates.getCoordinates()[0];
+      float lat = coordinates.getCoordinates()[1];
+      if (!coordinateBound(lon, lat)) {
+        throw new IllegalArgumentException("Coordinates are invalid");
+      }
+    }
+    return (Tweet) dao.create(tweet);
+  }
+
+  public boolean coordinateBound(float lon, float lat) {
+    lon = Math.abs(lon);
+    lat = Math.abs(lat);
+    return (lon <= 180 && lat <= 90);
+  }
+
+  @Override
+  public Tweet showTweet(String id, String[] fields) {
+    long longId;
+    try {
+      longId = Long.parseLong(id);
+    } catch (Exception i) {
+      throw new IllegalArgumentException("id param is invalid");
+    }
+    Tweet tweet = (Tweet) dao.findById(longId); //takes a while to run
+    Tweet workTweet = new Tweet();
+    Method[] methods = Tweet.class.getDeclaredMethods(); //all Tweet methods
+    Map<String,Method> methodNames = new HashMap<String,Method>();
+
+    for (Method m : methods) {
+      int lastBracket = m.toString().lastIndexOf('(');
+      String s = m.toString().substring(0,lastBracket);
+      int lastIndxDot = s.lastIndexOf('.');
+      String key = s.substring(lastIndxDot+1,lastBracket);
```

Comment:
m.getName() can save you this string manip. It may also be helpful to check if there are any fields to filter in the first place.